File: Python3/linkedlist.py
# Simple Linked List with Python

class Node:
    def __init__(self, value):
        self.value = value
        self.next_node = None


class LinkedList:
    def __init__(self):
        self.head = Node(None)
        self.tail = self.head  # references head node (same object)

    def append(self, node_val):
        next_node = Node(node_val)  # new node with given value as value
        temp = self.tail  # create temp reference to current tail
        temp.next_node = next_node
        self.tail = next_node

    def prepend(self, node_val):
        new_first = Node(node_val)
        temp = self.head.next_node
        self.head.next_node = new_first
        new_first.next_node = temp

    # inserted node will become nth element, nth element will become nth + 1
    def insert(self, node_val, index: int):
        new_node = Node(node_val)
        curr_node = self.head.next_node
        prev_node = None
        count = 0
        while count < index:
            prev_node = curr_node
            curr_node = curr_node.next_node
            count += 1
        prev_node.next_node = new_node
        new_node.next_node = curr_node

    def pop(self, index=-1):
        if index == 0:
            curr_node = self.head.next_node
            self.head.next_node = curr_node.next_node
        elif index > 0:
            count = 0
            curr_node = self.head.next_node
            prev_node = None
            next_node = None
            while count < index:
                prev_node = curr_node
                curr_node = curr_node.next_node
                next_node = curr_node.next_node
                count += 1
            prev_node.next_node = next_node
        # TODO: edit to allow for -2, -3, etc
        elif index == -1:
            curr_node = self.head
            prev_node = None
            while curr_node.next_node:
                prev_node = curr_node
                curr_node = curr_node.next_node
            prev_node.next_node = None
        else:
            pass

    def __iter__(self):
        self.curr_node = self.head
        return self

    def __next__(self):
        if self.curr_node.next_node:
            node_val = self.curr_node.next_node.value
            self.curr_node = self.curr_node.next_node
            return node_val
        raise StopIteration

    def __len__(self):
        return sum(1 for _ in self)

    def __str__(self):
        node = self.head
        out = ""
        while node.next_node:
            out += f"[{node.next_node.value}] -> "
            node = node.next_node
        return out[:-4]  # remove trailing arrow


def main():
    mylist = LinkedList()
    mylist.append(4)
    mylist.append(5)
    mylist.append(6)
    mylist.append(8)
    mylist.prepend(3)
    mylist.prepend(2)
    mylist.prepend(1)
    mylist.prepend(0)
    mylist.insert(7, 7)
    mylist.pop()
    mylist.pop(0)
    mylist.pop(5)

    print(mylist)

    for node in mylist:
        print(node)


if __name__ == "__main__":
    main()

File: JavaScript & TypeScript/linkedlist.js
// Simple linked list with JavaScript

class Node {
  constructor(value) {
    this.value = value
    this.nextNode = null
  }
}

class LinkedList {
  constructor() {
    this.head = new Node(null)
    this.tail = this.head
  }
  append(value) {
    let nextNode = new Node(value)
    let temp = this.tail
    temp.nextNode = nextNode
    this.tail = nextNode
  }
  prepend(value) {
    let newFirst = new Node(value)
    let temp = this.head.nextNode
    this.head.nextNode = newFirst
    newFirst.nextNode = temp
  }
  insert(value, index) {
    let newNode = new Node(value)
    let currNode = this.head.nextNode
    let prevNode = null
    let count = 0
    while (count < index) {
      prevNode = currNode
      currNode = currNode.nextNode
      count += 1
    }
    prevNode.nextNode = newNode
    newNode.nextNode = currNode
  }
  get(index) {
    let count = 0
    let currNode = this.head.nextNode
    while (count < index) {
      currNode = currNode.nextNode
      count += 1
    }
    return currNode.value
  }
  pop(index=-1) {
    if (index == 0) {
      let currNode = this.head.nextNode
      this.head.nextNode = currNode.nextNode
    }
    else if (index > 0) {
      let count = 0
      let currNode = this.head.nextNode
      let prevNode = null
      let nextNode = null
      while (count < index) {
        prevNode = currNode
        currNode = currNode.nextNode
        nextNode = currNode.nextNode
        count += 1
      }
      prevNode.nextNode = nextNode
    }
    else if (index == -1) {
      let currNode = this.head
      let prevNode = null
      while (currNode.nextNode) {
        prevNode = currNode
        currNode = currNode.nextNode
      }
      prevNode.nextNode = null
    }
    else { void(0) }
  }
  length() {
    let len = 0
    let currNode = this.head
    while (currNode.nextNode) {
      len += 1
      currNode = currNode.nextNode
    }
    return len
  }
  toString() {
    let currNode = this.head.nextNode
    let out = ""
    while (currNode) {
      out += "[" + String(currNode.value) + "] -> "
      currNode = currNode.nextNode
    }
    return out.slice(0, -4) // drop trailing arrow
  }
  [Symbol.iterator]() {
    let currNode = this.head
    return {
      next() {
        if (currNode.nextNode) {
          currNode = currNode.nextNode
          return { value: currNode.value, done: false }
        } else { 
          return { done: true }
        }
      }
    }
  }
}

mylist = new LinkedList()

mylist.append(4)
mylist.append(5)
mylist.append(6)
mylist.append(8)

mylist.prepend(3)
mylist.prepend(2)
mylist.prepend(1)
mylist.prepend(0)

mylist.insert(7, 7)

mylist.pop()
mylist.pop(0)
mylist.pop(5)

x = mylist.get(1)
console.log(x)

console.log(mylist.toString())
console.log(mylist.length())

for (let item of mylist) {
  console.log(item)
}